### ISSUE TITLE
possibly recording more nodes than we should with retainCoalescentOnly=F

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#487]
In simplification, if we've set `retainCoalescentOnly=F`, then we [include the `TSK_SIMPLIFY_KEEP_UNARY` flag](https://github.com/MesserLab/SLiM/blob/master/core/species.cpp#L4252):
```
		if (!retain_coalescent_only_) flags |= TSK_SIMPLIFY_KEEP_UNARY;
```
That [flag](https://tskit.dev/tskit/docs/latest/c-api.html#c.TSK_SIMPLIFY_KEEP_UNARY) does:
> By default simplify removes unary nodes (i.e., nodes with exactly one child) along the path from samples to root. If this option is specified such unary nodes will be preserved in the output. 

However, compare to [`TSK_SIMPLIFY_KEEP_UNARY_IN_INDIVIDUALS](https://tskit.dev/tskit/docs/latest/c-api.html#c.TSK_SIMPLIFY_KEEP_UNARY_IN_INDIVIDUALS):
> This acts like [TSK_SIMPLIFY_KEEP_UNARY](https://tskit.dev/tskit/docs/latest/c-api.html#c.TSK_SIMPLIFY_KEEP_UNARY) (and is mutually exclusive with that flag). It keeps unary nodes, but only if the unary node is referenced from an individual.

I suspect what we want here is actually the latter flag.

Indeed, in SLiMgui's help for `treeSeqRememberIndividuals`, it says:
> The behavior of simplification for individuals retained with permanent=F depends upon the value of the retainCoalescentOnly flag passed to initializeTreeSeq(); here we will discuss the behavior of that flag in detail.  First of all, genomes are always removed by simplification unless they are (a) part of the final generation (i.e., in a living individual when simplification occurs), (b) ancestral to the final generation, (c) a genome of a permanently remembered individual, or (d) ancestral to a permanently remembered individual.  If retainCoalescentOnly is T (the default), they are also always removed if they are not a branch point (i.e., a coalescent node or most recent common ancestor) in the tree sequence.  In some cases it may be useful to retain a genome and its associated individual when it is simply an intermediate node in the ancestry (i.e., in the middle of a branch).  This can be enabled by setting retainCoalescentOnly to F in your call to initializeTreeSeq().  In this case, ancestral genomes that are intermediate (“unary nodes”, in tskit parlance) and are within an individual that has been retained using the permanent=F flag here are kept, along with the retained individual itself. 

So, I think that
- currently if `retainCoalescentOnly=F` then we are retaining *all* nodes ancestral to the samples; while
- what we want to be doing is only retaining those nodes that are in Retained or Remembered individuals (that are ancestral to samples).

Note that if someone really did want to remember all ancestral nodes across all time they just need to do `rememberIndividuals(sim.subpopulations.individuals, permanent=F)` every generation.

We ought to verify this is the case, but I'm trying to not get distracted on my code review.

**Comment by petrelharp** [comment]
Note that this is not really so much a bug in that nothing is *wrong*, just we're recording more information than we ought to be.

**Comment by bhaller** [comment]
@hyanwong @petrelharp The actual proposed change to the code here is trivial, as I understand it.  I'd like you guys to confirm that it's really a change I should make, though.  Peter, you say "we ought to verify this is the case"; can you do something to verify that, perhaps?  (I don't know what that something ought to be, though.  :->)  And Yan, I think maybe these options came originally from requests that you made, if I recall correctly?  (Perhaps I don't recall correctly.  :->)  Does what is proposed here make sense to you, and do you think it is the right thing to do, even though it might break backward compatibility for some people who were actually getting the behavior they wanted from this flag?  Please let me know ASAP; I'm hoping to put SLiM 5 to bed in a couple of weeks, and this change ought to go into that, if it's ever going to happen.  Thanks!

**Comment by petrelharp** [comment]
Yes - I can confirm - here's what a tree looks like with `retainCoalescentOnly=F`:
```
         51     52 
          ┃      ┃ 
         50     49 
          ┃      ┃ 
         47     48 
          ┃      ┃ 
         46     45 
          ┃      ┃ 
         43     44 
       ┏━━┻━━┓   ┃ 
      40    42  41 
       ┃     ┃   ┃ 
      38    39  37 
       ┃     ┃   ┃ 
      36    35  34 
       ┃     ┃   ┃ 
      32    31  33 
       ┃     ┃   ┃ 
      30    28  29 
       ┃     ┃   ┃ 
      25    26  27 
       ┃     ┃   ┃ 
      24    22  23 
       ┃     ┃   ┃ 
      21    20  19 
    ┏━━┻━━┓  ┃   ┃ 
   16    18 17  15 
  ┏━┻━━┓  ┃  ┃   ┃ 
 11   12 14 10  13 
┏━╋━┓ ┏┻┓ ┃ ┏┻┓ ┏┻┓
0 2 9 1 3 5 7 8 4 6
```

Furthermore, confirmed in the help for `initializeTreeSeq`:
> The retainCoalescentOnly parameter controls how, exactly, simplification of the tree-sequence data is performed in SLiM (both for auto-simplification and for calls to treeSeqSimplify()).  More specifically, this parameter controls the behavior of simplification for individuals and genomes that have been “retained” by calling treeSeqRememberIndividuals() with the parameter permanent=F.  The default of retainCoalescentOnly=T helps to keep the number of retained individuals relatively small, which is helpful if your simulation regularly flags many individuals for retaining.  In this case, changing retainCoalescentOnly to F may dramatically increase memory usage and runtime, in a similar way to permanently remembering all the individuals.  See the documentation of treeSeqRememberIndividuals() for further discussion.

In other words - with `retainCoalescentOnly=F`, we want to be retaining unary nodes... but not *all* unary nodes, just those who we've remembered with permanent=F. But we used the wrong flag for that - easy mistake to make.

**Comment by bhaller** [comment]
OK, great, thanks @petrelharp!  I'll make the change, then, and document it in the release notes, but won't worry about the break in backward compatibility, since it will continue to work as documented.  :->  If somebody (@hyanwong?) wants the other behavior, that would presumably be a new flag with new documentation, and we'll cross that bridge when we come to it.  I'll make this change in `multichrom` as soon as I've finished what I'm working on right now.  :->

**Comment by petrelharp** [comment]
Great! FYI - I don't think we will need the other flag, because we can just remember *everyone* non-permanently, and that will be equivalent.

**Comment by bhaller** [comment]
OK, this is now fixed and pushed to the `multichrom` branch, which will be merged into `master` soon.  @petrelharp, perhaps the trees tests exercise this flag, so it might affect your testing work – just a heads-up.  Closing.
